### PR TITLE
Added competition info

### LINKED ISSUE #45: Store competition info
We should Store information regarding the programming competition on the DB.

Each competition should have a 1:1 relationship with its year's edition.

### PR DESCRIPTION
**Pull request by Naapperas**
Closes #45

### CODE REVIEW

File: resources/js/Pages/CRUD/Competition/Edit.vue [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    theme: competition.theme,
+});
+
+const submit = () => {
+    form.put(route("admin.competitions.update", competition));
+};
+</script>
+
+<template>
+    <CardLayout title="Criar Competição">
```
Comment (diogofonte) [inline resources/js/Pages/CRUD/Competition/Edit.vue]: Shouldn't it be "Edit Competition"?

**Comment by toni-santos** [comment]
Inputting an end date earlier than the start date is possible, this shouldn't be the case

**Review by pedronunes19 — DISMISSED** [review]

File: database/helpers.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * @param  \Illuminate\Database\Eloquent\Model|string  $model2
+ */
+function createManyToManyRelation($model1, $model2)
+{
+
+    // We need to ensure lexicographical order because Laravel
+    $models = [$model1, $model2];
+    sort($models, SORT_STRING);
+    $model1 = $models[0];
+    $model2 = $models[1];
```
Comment (ttoino) [inline database/helpers.php]: You can use array destructuring for this

File: database/helpers.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    if (is_string($model1)) {
+        $model1 = new $model1;
+    }
+
+    if (is_string($model2)) {
+        $model2 = new $model2;
+    }
```
Comment (ttoino) [inline database/helpers.php:28]: Why is this needed?

File: resources/js/Pages/CRUD/Competition/Create.vue [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            <select v-model="form.edition_id" required class="self-stretch">
+                <option value="" disabled selected hidden>Edição</option>
+                <option
+                    v-for="edition in $props.with.editions"
+                    :key="edition.id"
+                    :value="edition.id"
+                >
+                    {{ edition.name }}
+                </option>
+            </select>
```
Comment (ttoino) [inline resources/js/Pages/CRUD/Competition/Create.vue]: We should use the new TextInput[type=select]

File: resources/js/Pages/CRUD/Competition/Edit.vue [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            <select v-model="form.edition_id" required class="self-stretch">
+                <option value="" disabled selected hidden>Edição</option>
+                <option
+                    v-for="edition in $props.with.editions"
+                    :key="edition.id"
+                    :value="edition.id"
+                >
+                    {{ edition.name }}
+                </option>
+            </select>
```
Comment (ttoino) [inline resources/js/Pages/CRUD/Competition/Edit.vue]: Same as above

File: resources/js/Pages/CRUD/CompetitionTeam/Create.vue [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+            <select v-model="form.competition_id" required class="self-stretch">
+                <option value="" disabled selected hidden>Competição</option>
+                <option
+                    v-for="competition in $props.with.competitions"
+                    :key="competition.id"
+                    :value="competition.id"
+                >
+                    {{ competition.edition?.name ?? competition.id }}
+                </option>
+            </select>
```
Comment (ttoino) [inline resources/js/Pages/CRUD/CompetitionTeam/Create.vue]: Same as above

File: resources/js/Pages/CRUD/CompetitionTeam/Edit.vue [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+            <select v-model="form.competition_id" required class="self-stretch">
+                <option value="" disabled selected hidden>Competição</option>
+                <option
+                    v-for="competition in $props.with.competitions"
+                    :key="competition.id"
+                    :value="competition.id"
+                >
+                    {{ competition.id }}
+                </option>
+            </select>
```
Comment (ttoino) [inline resources/js/Pages/CRUD/CompetitionTeam/Edit.vue]: Same as above

File: resources/js/Pages/CRUD/CompetitionTeam/Index.vue [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        name="competitionTeams"
+    >
+        <template #heading>Equipas na Competição de Programação</template>
+
+        <template #header>
+            <HeaderRow>
+                <Header sort-by="name">Edição</Header>
+                <Header>Nome</Header>
+                <Header>Nª Membros</Header>
+                <Header>Pontos</Header>
```
Comment (ttoino) [inline resources/js/Pages/CRUD/CompetitionTeam/Index.vue]: Careful with the Header attributes

File: resources/js/Types/Competition.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import type Model from "@/Types/Model";
+import type Edition from "@/Types/Edition";
+import type CompetitionTeam from "@/Types/Competition";
+
+export default interface Competition extends Model {
+    theme: string;
+    date_start: Date;
+    date_end: Date;
```
Comment (ttoino) [inline resources/js/Types/Competition.ts]: This should not be Date, the data that comes from the server is just a string

**Review by ttoino — CHANGES_REQUESTED** [review]

**Review by toni-santos — DISMISSED** [review]

File: database/helpers.php (hunk 3 above)
Comment (Naapperas) [reply database/helpers.php:28]: I just took this from the Laravel source code, even though I believe we will never _not_ use the `model::class` syntax this is a safety measure.

File: resources/js/Pages/CRUD/Competition/Create.vue (hunk 4 above)
Comment (Naapperas) [reply resources/js/Pages/CRUD/Competition/Create.vue]: This was made before that change, nice catch.

**Review by ttoino — APPROVED** [review]

**Review by toni-santos — APPROVED** [review]

**Pull request merged**